src/pushsource/_impl/backend/koji_source.py

```diff
@@ -504,6 +514,108 @@ def _push_items_from_container_build(self, nvr, meta):
 
         return out
 
+    def _push_items_from_vmi_build(self, nvr, meta):
+        LOG.debug("Looking for virtual machine images on %s, %s", nvr, meta)
+
+        if not meta:
+            message = "Virtual machine image build not found in koji: %s" % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        if meta.get("container_koji_task_id"):
+            message = "The build %s is for container images." % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        # VMI has the {'typeinfo': {'image':{}}} on extra
+        extra = meta.get("extra") or {}
+        typeinfo = extra.get("typeinfo") or {}
+        image = typeinfo.get("image")
+
+        if image is None:
+            message = "Build %s not recognized as a virtual machine image build" % nvr
+            LOG.debug(message)
+            raise ValueError(message)
+
+        build_id = meta["id"]
+
+        archives = self._get_archives(build_id)
+
+        # Supported types of VMI archives
+        ami_types = ["raw", "raw-xz"]
+        azure_types = ["vhd", "vhd-compressed"]
+        cloud_types = {}
+
+        for k in ami_types:
+            cloud_types.setdefault(k, AmiPushItem)
+
+        for k in azure_types:
+            cloud_types.setdefault(k, VHDPushItem)
+
+        # Collect all known VM image archives
+        vmi_archives = [
+            elem
+            for elem in archives
+            if elem.get("btype") == "image"
+            and elem.get("type_name") in cloud_types.keys()
+        ]
+
+        out = []
+
+        # Generate the respective PushItem type for each archive
+        for archive in vmi_archives:
+            path = self._pathinfo.typedir(meta, archive["btype"])
+            item_src = os.path.join(path, archive["filename"])
+
+            extra = archive.get("extra") or {}
+            image = extra.get("image") or {}
+            arch = image.get("arch")
+
+            klass = cloud_types.get(archive["type_name"]) or VMIPushItem
+
+            rel_klass = klass._RELEASE_TYPE
+
+            # The given name will be composed by the product name and cloud
+            # marketpalce name, like "rhel-azure" or "rhel-sap-azure".
+            # We just want to get the product name without the cloud name,
+            # like "rhel" and "rhel-sap" in our examples respectively.
+            splitted_name = meta["name"].split("-")
+            if len(splitted_name) > 1:
+                product = "-".join(splitted_name[:-1])
+            else:
+                product = splitted_name[0]
+
+            # Try to get the "int" part of release for respin
```

Comment:
Hmm, I suppose this won't crash now, but does it make sense to use this value as "respin" really?

In the example given previously, the value would be 20230201, that doesn't seem like it fits what a "respin" is documented to be.  The doc string explicitly says "Respin count. 0 for original build of an image, incremented by one for each rebuild", but the image wasn't rebuilt twenty million times.

If there isn't a reasonable value here which actually maps to "respin", I think it'd be better to just set it as 0 always, or even update the model to allow respin to be None.  I note that the full release info is still present in the KojiBuildInfo attached to the item, so leaving this out won't make anything technically impossible in your code using the items, as if you wanted to make use of the release in some way you'll still be able to do that.

Reply:
I agree with you.

I've changed it to pick from "release" if it's an integer or set it as 0 if not.